crates/bevy_pbr/src/volumetric_fog/volumetric_fog.wgsl

```diff
@@ -0,0 +1,214 @@
+    return FRAC_4_PI * (1.0 - g * g) / (denom * sqrt(denom));
+}
+
+@fragment
+fn fragment(in: FullscreenVertexOutput) -> @location(0) vec4<f32> {
+    // Unpack the `volumetric_fog` settings.
+    let fog_color = volumetric_fog.fog_color;
+    let ambient_color = volumetric_fog.ambient_color;
+    let ambient_intensity = volumetric_fog.ambient_intensity;
+    let step_count = volumetric_fog.step_count;
+    let max_depth = volumetric_fog.max_depth;
+    let absorption = volumetric_fog.absorption;
+    let scattering = volumetric_fog.scattering;
+    let density = volumetric_fog.density;
+    let light_tint = volumetric_fog.light_tint;
+    let light_intensity = volumetric_fog.light_intensity;
+
+    let exposure = view.exposure;
+
+    // Sample the depth. If this is multisample, just use sample 0; this is
+    // approximate but good enough.
+    let frag_coord = in.position;
+    let depth = textureLoad(depth_texture, vec2<i32>(frag_coord.xy), 0);
+
+    // Starting at the end depth, which we got above, figure out how long the
+    // ray we want to trace is and the length of each increment.
+    let end_depth = min(
+        volumetric_fog.max_depth,
+        -position_ndc_to_view(frag_coord_to_ndc(vec4(in.position.xy, depth, 1.0))).z
+    );
+    let step_size = end_depth / f32(step_count);
+
+    let directional_light_count = lights.n_directional_lights;
+
+    // Calculate the ray origin (`Ro`) and the ray direction (`Rd`) in NDC,
+    // view, and world coordinates.
+    let Rd_ndc = vec3(frag_coord_to_ndc(in.position).xy, 1.0);
+    let Rd_view = normalize(position_ndc_to_view(Rd_ndc));
+    let Ro_world = view.world_position;
+    let Rd_world = normalize(position_ndc_to_world(Rd_ndc) - Ro_world);
+
+    // Use Beer's law [1] [2] to calculate the maximum amount of light that each
+    // directional light could contribute, and modulate that value by the light
+    // tint and fog color. (The actual value will in turn be modulated by the
+    // phase according to the Henyey-Greenstein formula.)
+    //
+    // We use a bit of a hack here. Conceptually, directional lights are
+    // infinitely far away. But, if we modeled exactly that, then directional
+    // lights would never contribute any light to the fog, because an
+    // infinitely-far directional light combined with an infinite amount of fog
+    // would result in complete absorption of the light. So instead we pretend
+    // that the directional light is `max_depth` units away and do the
+    // calculation in those terms. This simplification also allows us to avoid
+    // shooting secondary rays toward the directional light, which
+    // asymptotically improves performance.
+    //
+    // [1]: https://www.scratchapixel.com/lessons/3d-basic-rendering/volume-rendering-for-developers/intro-volume-rendering.html
+    //
+    // [2]: https://en.wikipedia.org/wiki/Beer%E2%80%93Lambert_law
+    let light_attenuation = exp(-density * max_depth * (absorption + scattering));
```

Comment:
I think I get the basic idea of this hack, but I'm still struggling a little bit to understand exactly what this attenuation represents; based on the "not needing to shoot secondary rays toward the directional light" bit, I'm imagining that this is supposed to represent, for each point in the raymarch, the light attenuation of traveling from the directional light to the sample point? If so, I think that could be a little clearer.

(That would also clear up my confusion over the line following this, where the density, step size, and scattering coefficient appear once again — because they are for the attenuation on the journey to the camera from the sample point, I think.)

Reply:
Yes. I reworded the comment.